src/librbd/cache/pwl/AbstractWriteLog.cc

```diff
@@ -637,6 +637,19 @@ void AbstractWriteLog<I>::shut_down(Context *on_finish) {
       }
       update_image_cache_state(next_ctx);
     });
+  ctx = new LambdaContext(
+    [this, ctx](int r) {
+      ldout(m_image_ctx.cct, 6) << "complete in flight operations" << dendl;
+      m_work_queue.queue(ctx, r);
+    });
+  ctx = new LambdaContext(
+    [this, ctx](int r) {
+      Context *next_ctx = override_ctx(r, ctx);
+      ldout(m_image_ctx.cct, 6) << "waiting for in flight operations" << dendl;
+      // Wait for in progress IOs to complete
+      next_ctx = util::create_async_context_callback(m_image_ctx, next_ctx);
+      m_async_op_tracker.wait_for_ops(next_ctx);
```

Comment:
dead lock is due to op finish in advance(call finish_op) but function not finish, it want get lock, see https://github.com/ceph/ceph/pull/42950#issuecomment-920611528, and https://github.com/ceph/ceph/pull/42950/commits/7da4d71ead2bf1d7ca49fc2b60d7e12f2df9de47 this commit fix the dead lock, move finish_op() to the end.

Here I change the order just want the `remove_pool_file()` run in `m_work_queue`, Otherwise, if it may runs in `aio_thread` from `flush_dirty_entries()` call back,  then bdev->close() calles _aio_stop(), exec aio_thread.join(), cause assert. Thread can't join itself. 

+@ideepika  moving flight operations from earlier position  has nothing to do with the assert bug of this commit. But I think about `flush_dirty_entries()` will call `wake up()`, then may call  `m_async_op_tracker.start_op()`. So would it be better to put `wait_for_ops` after `flush_dirty_entries()`?

Reply:
makes sense to me, thanks for the update!

Reply:
I am still not quite sure I understand if we need here  `util::create_async_context_callback(m_image_ctx, next_ctx)`, i.e. pass the contexct to `m_image_ctx.op_work_queue`, which is passed after this to `m_work_queue.queue`.

Wouldn't we have the same effect if just change:
```
next_ctx = util::create_async_context_callback(m_image_ctx, next_ctx);
```
to
```
next_ctx = util::create_async_context_callback(m_work_queue, next_ctx);
```
And will not add
```
ctx = new LambdaContext(
   [this, ctx](int r) {
     m_work_queue.queue(ctx, r);
 });
 ```

Reply:
No, it's my problem. I haven't used it like this, so I don't understand it. I will try to read the code to understand it and modify it as you suggest.